src/test/scala/org/broadinstitute/dsde/firecloud/dataaccess/HttpGoogleServicesDAOSpec.scala

```diff
@@ -63,4 +65,86 @@ class HttpGoogleServicesDAOSpec extends FlatSpec with Matchers {
     objectMetadata.name should equal("test-composite-object")
     objectMetadata.md5Hash should equal(None)
   }
+
+  {
+    val headers = List[AnyRef]("header 1", "header 2", "header 3").asJava
+
+    val row1 = List[AnyRef]("proj 1", "free", "trial").asJava
+    val row2 = List[AnyRef]("proj 2", "firecloud", "credits").asJava
+    val row3 = List[AnyRef]("proj 3", "test", "data").asJava
+
+    val cells1 = List(headers, row1, row2, row3)
+
+    it should "have no changes if the data didn't change" in {
+      check(
+        newContent = cells1,
+        existingContent = cells1,
+        expectedOutput = cells1
+      )
+    }
+
+    it should "not explode if there is no data (only headers)" in {
+      check(
+        newContent = List(headers),
+        existingContent = List(headers),
+        expectedOutput = List(headers)
+      )
+    }
+
+    it should "not explode or delete data if the update has no data" in {
+      check(
+        newContent = List(headers),
+        existingContent = cells1,
+        expectedOutput = cells1
+      )
+    }
+
+    it should "add a new row without modifying existing ones" in {
+      check(
+        newContent = List(headers, row1, row2, row3),
+        existingContent = List(headers, row1, row2),
+        expectedOutput = List(headers, row1, row2, row3)
+      )
+    }
+
+    it should "output the union of the rows if new is missing a row compared to existing" in {
+      check(
+        newContent = List(headers, row2, row3),
+        existingContent = List(headers, row1, row2),
+        expectedOutput = List(headers, row1, row2, row3)
+      )
+    }
+
+    it should "locate a row in existing that moved based on its first column, and replace its remaining columns with new" in {
+
+      val newRow1 = List[AnyRef](row1.get(0), "new data first col!", "new data second col!").asJava
+
+      check(
+        newContent = List(headers, row2, row3, newRow1),
+        existingContent = List(headers, row1, row2, row3),
+        expectedOutput = List(headers, newRow1, row2, row3) // Row 1 is still first, but has the new data
+      )
+    }
+
+    it should "preserve order of rows if all the updates are out of order" in {
+      check(
+        newContent = List(headers, row2, row3, row1),
+        existingContent = List(headers, row1, row2, row3),
+        expectedOutput = List(headers, row1, row2, row3)
+      )
+    }
+
+    def check(newContent: List[java.util.List[AnyRef]], existingContent: List[java.util.List[AnyRef]], expectedOutput: List[java.util.List[AnyRef]]): Unit = {
+      // https://stackoverflow.com/a/24375762/818054
+      val updatePreservingOrder = PrivateMethod('updatePreservingOrder)
```

Comment:
If people don't like this (and I can see why they wouldn't), happy to just make `updatePreservingOrder` `public`

Reply:
Cool! I didn't know this was possible.